src/main/java/mezz/jei/ingredients/IngredientListElementComparator.java

```diff
@@ -1,41 +1,483 @@
 package mezz.jei.ingredients;
 
-import java.util.Comparator;
-
 import mezz.jei.config.Constants;
 import mezz.jei.gui.ingredients.IIngredientListElement;
 import net.minecraft.item.ItemStack;
+import net.minecraft.item.ItemTool;
+import mezz.jei.config.Config;
+import mezz.jei.util.Translator;
+import net.minecraft.entity.SharedMonsterAttributes;
+import net.minecraft.entity.ai.attributes.AttributeModifier;
+import net.minecraft.inventory.EntityEquipmentSlot;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemArmor;
+import net.minecraft.item.ItemHoe;
+import net.minecraft.item.ItemShears;
+import scala.Int;
+import java.util.*;
+
+import com.google.common.collect.Multimap;
 
 public final class IngredientListElementComparator implements Comparator<IIngredientListElement> {
 	public static final IngredientListElementComparator INSTANCE = new IngredientListElementComparator();
 
+	public static ArrayList<SortEntry> entries = new ArrayList<SortEntry>();
+	public static ArrayList<SortEntry> list = new ArrayList<SortEntry>();
+	private static boolean areDefaultEntriesLoaded = false;
+
 	private IngredientListElementComparator() {
 
 	}
 
+	public static class SortEntry {
+		public String name;
+		// External interfaces only have access to ItemStacks, sorry fluids.
+		public Comparator<ItemStack> itemStackComparator;
+		// This is the internal comparator, fluids supported here.
+		public Comparator<IIngredientListElement> ingredientComparator;
+
+		public SortEntry(String name, Comparator<ItemStack> comparator) {
+			this.name = name;
+			this.itemStackComparator = comparator;
+		}
+
+		public SortEntry(String name) {
+			this.name = name;
+		}
+
+		public String getLocalisedName() {
+			return Translator.translateToLocal(name);
+		}
+
+		public String getTooltip() {
+			String tipname = name + ".tip";
+			String tip = Translator.translateToLocal(tipname);
+			return !tip.equals(tipname) ? tip : null;
+		}
+	}
+
 	@Override
 	public int compare(IIngredientListElement o1, IIngredientListElement o2) {
-		final String modName1 = o1.getModNameForSorting();
-		final String modName2 = o2.getModNameForSorting();
-
-		if (modName1.equals(modName2)) {
-			boolean isItemStack1 = (o1.getIngredient() instanceof ItemStack);
-			boolean isItemStack2 = (o2.getIngredient() instanceof ItemStack);
-			if (isItemStack1 && !isItemStack2) {
-				return -1;
-			} else if (!isItemStack1 && isItemStack2) {
-				return 1;
-			}
-
-			final int orderIndex1 = o1.getOrderIndex();
-			final int orderIndex2 = o2.getOrderIndex();
-			return Integer.compare(orderIndex1, orderIndex2);
-		} else if (modName1.equals(Constants.MINECRAFT_NAME)) {
-			return -1;
-		} else if (modName2.equals(Constants.MINECRAFT_NAME)) {
-			return 1;
-		} else {
-			return modName1.compareTo(modName2);
+		if (list.size() == 0) {
```

Comment:
We should not make this a singleton class. 
That way `list` can be passed into the constructor and we can be confident that anywhere we can access an instance of it, things are set correctly already.

Reply:
Ok.

Reply:
I made the empty constructor default to the config value.  There is also a constructor that takes a string and ignores the config.